File: repertoire/data-integration/providers/transform_plugin/JsonExtract.swift
// Transform Plugin Provider: json_extract
// Extract values from JSON strings at a specified path.
// See Architecture doc for transform plugin interface contract.

import Foundation

public final class JsonExtractTransformProvider {
    public static let providerId = "json_extract"
    public static let pluginType = "transform_plugin"

    public init() {}

    public func transform(value: Any, config: TransformConfig) throws -> Any {
        if value is NSNull { return NSNull() }

        let path = (config.options["path"] as? String) ?? "$"
        let defaultValue = config.options["default"]
        let hasDefault = config.options.keys.contains("default")

        let parsed: Any

        if let str = value as? String {
            guard let data = str.data(using: .utf8),
                  let jsonObj = try? JSONSerialization.jsonObject(with: data) else {
                throw TransformError.invalidCast("Invalid JSON input")
            }
            parsed = jsonObj
        } else {
            parsed = value
        }

        if let result = navigatePath(root: parsed, path: path) {
            return result
        }

        if hasDefault {
            return defaultValue ?? NSNull()
        }

        return NSNull()
    }

    private func navigatePath(root: Any, path: String) -> Any? {
        if path == "$" || path.isEmpty {
            return root
        }

        var normalized = path
        if normalized.hasPrefix("$.") {
            normalized = String(normalized.dropFirst(2))
        } else if normalized.hasPrefix("$") {
            normalized = String(normalized.dropFirst(1))
        }

        let segments = tokenizePath(normalized)
        var current: Any = root

        for segment in segments {
            // Array index access [N]
            if segment.hasPrefix("[") && segment.hasSuffix("]") {
                let indexStr = String(segment.dropFirst().dropLast())
                if indexStr == "*" {
                    // Wildcard: return all elements
                    if let arr = current as? [Any] { return arr }
                    if let dict = current as? [String: Any] { return Array(dict.values) }
                    return nil
                }
                guard let index = Int(indexStr), let arr = current as? [Any],
                      index >= 0, index < arr.count else {
                    return nil
                }
                current = arr[index]
            } else if segment == "*" {
                if let arr = current as? [Any] { return arr }
                if let dict = current as? [String: Any] { return Array(dict.values) }
                return nil
            } else {
                // Property access
                guard let dict = current as? [String: Any], let val = dict[segment] else {
                    return nil
                }
                current = val
            }
        }

        return current
    }

    private func tokenizePath(_ path: String) -> [String] {
        var segments: [String] = []
        var current = ""
        var i = path.startIndex

        while i < path.endIndex {
            let ch = path[i]

            if ch == "." {
                if !current.isEmpty {
                    segments.append(current)
                    current = ""
                }
                i = path.index(after: i)
            } else if ch == "[" {
                if !current.isEmpty {
                    segments.append(current)
                    current = ""
                }
                // Find closing bracket
                guard let closeIdx = path[i...].firstIndex(of: "]") else {
                    current.append(ch)
                    i = path.index(after: i)
                    continue
                }
                let bracketContent = String(path[i...closeIdx])
                // Check for string keys: ['key'] or ["key"]
                let inner = String(bracketContent.dropFirst().dropLast())
                if (inner.hasPrefix("'") && inner.hasSuffix("'"))
                    || (inner.hasPrefix("\"") && inner.hasSuffix("\"")) {
                    segments.append(String(inner.dropFirst().dropLast()))
                } else {
                    segments.append(bracketContent)
                }
                i = path.index(after: closeIdx)
            } else {
                current.append(ch)
                i = path.index(after: i)
            }
        }

        if !current.isEmpty {
            segments.append(current)
        }

        return segments
    }

    public func inputType() -> TypeSpec {
        return TypeSpec(type: "string", nullable: true)
    }

    public func outputType() -> TypeSpec {
        return TypeSpec(type: "any", nullable: true)
    }
}

File: repertoire/data-integration/providers/transform_plugin/json_extract.ts
// Transform Plugin Provider: json_extract
// Extract values from JSON strings at a specified path.
// See Architecture doc for transform plugin interface contract.

export const PROVIDER_ID = 'json_extract';
export const PLUGIN_TYPE = 'transform_plugin';

export interface TransformConfig {
  options?: Record<string, unknown>;
}

export interface TypeSpec {
  type: string;
  nullable?: boolean;
}

export class JsonExtractTransformProvider {
  transform(value: unknown, config: TransformConfig): unknown {
    if (value === null || value === undefined) {
      return null;
    }

    const path = (config.options?.path as string) ?? '$';
    const defaultValue = config.options?.default;
    const hasDefault = config.options !== undefined && 'default' in config.options;

    let parsed: unknown;

    if (typeof value === 'string') {
      try {
        parsed = JSON.parse(value);
      } catch (err) {
        throw new Error(`Invalid JSON input: ${(err as Error).message}`);
      }
    } else {
      parsed = value;
    }

    try {
      const result = this.navigatePath(parsed, path);
      if (result === undefined) {
        return hasDefault ? defaultValue : null;
      }
      return result;
    } catch {
      if (hasDefault) return defaultValue;
      throw new Error(`Cannot extract path "${path}" from JSON`);
    }
  }

  private navigatePath(obj: unknown, path: string): unknown {
    if (path === '$' || path === '') {
      return obj;
    }

    // Normalize path: strip leading $. or $
    let normalized = path;
    if (normalized.startsWith('$.')) {
      normalized = normalized.substring(2);
    } else if (normalized.startsWith('$')) {
      normalized = normalized.substring(1);
    }

    // Tokenize path segments: support dot notation and bracket notation
    const segments = this.tokenizePath(normalized);

    let current: unknown = obj;

    for (const segment of segments) {
      if (current === null || current === undefined) {
        return undefined;
      }

      // Array index access
      const indexMatch = segment.match(/^\[(\d+)\]$/);
      if (indexMatch) {
        const idx = parseInt(indexMatch[1], 10);
        if (Array.isArray(current)) {
          current = current[idx];
        } else {
          return undefined;
        }
        continue;
      }

      // Wildcard: collect all values
      if (segment === '*' || segment === '[*]') {
        if (Array.isArray(current)) {
          return current;
        }
        if (typeof current === 'object' && current !== null) {
          return Object.values(current);
        }
        return undefined;
      }

      // Property access
      if (typeof current === 'object' && current !== null) {
        current = (current as Record<string, unknown>)[segment];
      } else {
        return undefined;
      }
    }

    return current;
  }

  private tokenizePath(path: string): string[] {
    const segments: string[] = [];
    let i = 0;

    while (i < path.length) {
      if (path[i] === '.') {
        i++;
        continue;
      }

      if (path[i] === '[') {
        const close = path.indexOf(']', i);
        if (close === -1) {
          throw new Error(`Unclosed bracket in path at position ${i}`);
        }
        let key = path.substring(i, close + 1);
        // String key in brackets: ['key'] or ["key"]
        const strMatch = key.match(/^\[['"](.+)['"]\]$/);
        if (strMatch) {
          segments.push(strMatch[1]);
        } else {
          segments.push(key);
        }
        i = close + 1;
        continue;
      }

      // Read until next dot or bracket
      let end = i;
      while (end < path.length && path[end] !== '.' && path[end] !== '[') {
        end++;
      }
      segments.push(path.substring(i, end));
      i = end;
    }

    return segments;
  }

  inputType(): TypeSpec {
    return { type: 'string', nullable: true };
  }

  outputType(): TypeSpec {
    return { type: 'any', nullable: true };
  }
}

export default JsonExtractTransformProvider;
